src/java.base/share/classes/javax/crypto/KDF.java

```
@@ -0,0 +1,669 @@
+        }
+    }
+
+    /**
+     * Use the firstSpi as the chosen KDFSpi and set the fields accordingly
+     */
+    private void useFirstSpi() {
+        if (checkSpiNonNull(theOne)) return;
+
+        synchronized (lock) {
+            theOne = candidate;
+        }
+    }
+
+    /**
+     * Selects the provider which supports the passed {@code algorithm} and
+     * {@code derivationSpec} values, and assigns the global spi and provider
+     * variables if they have not been assigned yet.
+     * <p>
+     * If the spi has already been set, it will just return the result.
+     */
+    private Object chooseProvider(String algorithm,
+                                  AlgorithmParameterSpec derivationSpec)
+            throws InvalidAlgorithmParameterException,
+                   NoSuchAlgorithmException {
+
+        boolean isDeriveData = (algorithm == null);
+
+        if (checkSpiNonNull(theOne)) {
+            return (isDeriveData) ?
+                           theOne.spi().engineDeriveData(derivationSpec) :
+                           theOne.spi().engineDeriveKey(algorithm, derivationSpec);
+        }
+
+        synchronized (lock) {
+            Exception lastException = null;
+            if (!checkSpiNonNull(candidate)) {
+                throw new RuntimeException("Unexpected Error: candidate is null!");
+            }
+            Delegate currOne = candidate;
+            try {
+                while (true) {
+                    try {
+                        Object result = (isDeriveData) ?
+                                currOne.spi().engineDeriveData(derivationSpec) :
+                                currOne.spi().engineDeriveKey(algorithm,
+                                        derivationSpec);
+                        // found a working KDFSpi
+                        this.theOne = currOne;
+                        return result;
+                    } catch (Exception e) {
+                        if (lastException == null) {
```

Add some debug output since `e` can be silently discarded.